File: Server_Root/Server/RoomServer/Factory.cpp
#include "stdafx.h"
#include "Factory.h"
#include "Entity\Entities\Character\Character.h"
//#include "Entity\Entities\Character\CharacterAI.h"
#include "MasterData\MasterDataManager.h"
#include "MasterData\Behavior.h"
#include "MasterData\Character.h"
#include "MasterData\Skill.h"
#include "MasterData\State.h"
#include "Behavior\Behaviors\General.h"
#include "Behavior\Behaviors\Move.h"
#include "Behavior\Behaviors\Dash.h"
#include "Behavior\Behaviors\Die.h"
#include "Skill\Skills\ContinueSkill.h"
#include "Skill\Skills\FireSkill.h"
#include "State\States\Acceleration.h"
#include "State\States\Shield.h"
#include "State\States\FireBehavior.h"
#include "State\States\GeneralState.h"
#include "State\States\ChallengerDisturbing.h"
#include "Entity\Entities\Projectile\Projectile.h"

Factory::Factory()
{
}

Factory::~Factory()
{
}

ISkill* Factory::CreateSkill(BaeGameRoom* pGameRoom, IEntity* pEntity, int nMasterDataID)
{
	MasterData::Skill* pSkill = NULL;
	MasterDataManager::Instance()->GetData<MasterData::Skill>(nMasterDataID, pSkill);

	string strClassName = pSkill->m_strClassName;

	if (ContinueSkill::NAME.compare(strClassName) == 0)
	{
		return new ContinueSkill(pGameRoom, pEntity, nMasterDataID);
	}
	else if (FireSkill::NAME.compare(strClassName) == 0)
	{
		return new FireSkill(pGameRoom, pEntity, nMasterDataID);
	}

	return NULL;
}

IBehavior* Factory::CreateBehavior(BaeGameRoom* pGameRoom, IEntity* pEntity, int nMasterDataID)
{
	MasterData::Behavior* pMasterBehavior = NULL;
	MasterDataManager::Instance()->GetData<MasterData::Behavior>(nMasterDataID, pMasterBehavior);

	string strClassName = pMasterBehavior->m_strClassName;

	if (General::NAME.compare(strClassName) == 0)
	{
		return new General(pGameRoom, pEntity, nMasterDataID);
	}
	else if (Move::NAME.compare(strClassName) == 0)
	{
		return new Move(pGameRoom, pEntity, nMasterDataID);
	}
	else if (Dash::NAME.compare(strClassName) == 0)
	{
		return new Dash(pGameRoom, pEntity, nMasterDataID);
	}
	else if (Die::NAME.compare(strClassName) == 0)
	{
		return new Die(pGameRoom, pEntity, nMasterDataID);
	}

	return NULL;
}

IState* Factory::CreateState(BaeGameRoom* pGameRoom, IEntity* pEntity, int nMasterDataID, long long lStartTime)
{
	MasterData::State* pMasterState = NULL;
	MasterDataManager::Instance()->GetData<MasterData::State>(nMasterDataID, pMasterState);

	string strClassName = pMasterState->m_strClassName;

	if (Acceleration::NAME.compare(strClassName) == 0)
	{
		return new Acceleration(pGameRoom, pEntity, nMasterDataID, lStartTime);
	}
	else if (Shield::NAME.compare(strClassName) == 0)
	{
		return new Shield(pGameRoom, pEntity, nMasterDataID, lStartTime);
	}
	else if (FireBehavior::NAME.compare(strClassName) == 0)
	{
		return new FireBehavior(pGameRoom, pEntity, nMasterDataID, lStartTime);
	}
	else if (GeneralState::NAME.compare(strClassName) == 0)
	{
		return new GeneralState(pGameRoom, pEntity, nMasterDataID, lStartTime);
	}
	else if (ChallengerDisturbing::NAME.compare(strClassName) == 0)
	{
		return new ChallengerDisturbing(pGameRoom, pEntity, nMasterDataID, lStartTime);
	}

	return NULL;
}

Character* Factory::CreateCharacter(BaeGameRoom* pGameRoom, int nID, int nMasterDataID, Character::Role role)
{
	return new Character(pGameRoom, nID, nMasterDataID, role);
}

//CharacterAI* Factory::CreateCharacterAI(BaeGameRoom* pGameRoom, int nID, int nMasterDataID)
//{
//	return new CharacterAI(pGameRoom, nID, nMasterDataID);
//}

Projectile* Factory::CreateProjectile(BaeGameRoom* pGameRoom, int nProjectorID, int nID, int nMasterDataID)
{
	return new Projectile(pGameRoom, nProjectorID, nID, nMasterDataID);
}

File: Client_Root/Client/Assets/Scripts/Common/Factory.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Factory : MonoSingleton<Factory>
{
//    public ISkill         CreateSkill(BaeGameRoom* pGameRoom, IEntity* pEntity, int nMasterDataID)
//    {
//    }

    public IBehavior CreateBehavior(int nMasterDataID)
    {
        MasterData.Behavior masterData = null;
        MasterDataManager.Instance.GetData<MasterData.Behavior>(nMasterDataID, ref masterData);

        if (masterData.m_strClassName == "Move")
        {
            return new Behavior.Move();
        }
        else if (masterData.m_strClassName == "Rotation")
        {
            return new Behavior.Rotation();
        }
        else if (masterData.m_strClassName == "Idle")
        {
            return new Behavior.Idle();
        }
        else if (masterData.m_strClassName == "Jump")
        {
            return new Behavior.Jump();
        }
        else if (masterData.m_strClassName == "Cast")
        {
            return new Behavior.Cast();
        }
        else if (masterData.m_strClassName == "Die")
        {
            return new Behavior.Die();
        }

        return null;
    }

    public IState CreateState(int nMasterDataID)
    {
        MasterData.State masterData = null;
        MasterDataManager.Instance.GetData<MasterData.State>(nMasterDataID, ref masterData);

        if (masterData.m_strClassName == "ChallengerDisturbing")
        {
            return ObjectPool.instance.GetObject<State.ChallengerDisturbing>();
        }
        else if (masterData.m_strClassName == "Faint")
        {
            return ObjectPool.instance.GetObject<State.Faint>();
        }
        else if (masterData.m_strClassName == "General")
        {
            return ObjectPool.instance.GetObject<State.General>();
        }
        else if (masterData.m_strClassName == "Boost")
        {
            return ObjectPool.instance.GetObject<State.Boost>();
        }
        else if (masterData.m_strClassName == "Shield")
        {
            return ObjectPool.instance.GetObject<State.Shield>();
        }
		else if (masterData.m_strClassName == "Flicker")
        {
            return ObjectPool.instance.GetObject<State.Flicker>();
        }

        return null;
    }

    public Character CreateCharacter(int nMasterDataID)
    {
        GameObject goCharacter = ObjectPool.Instance.GetGameObject("CharacterModel/Character");
        Character character = goCharacter.GetComponent<Character>();

        return character;
    }

    public Projectile CreateProjectile(int nMasterDataID)
    {
        GameObject goProjectile = ObjectPool.Instance.GetGameObject("ProjectileModel/Projectile");
        Projectile projectile = goProjectile.GetComponent<Projectile>();

        return projectile;
    }

    public GameItem CreateGameItem()
    {
        GameObject goGameItem = ObjectPool.Instance.GetGameObject("ItemModel/GameItem");
        GameItem gameItem = goGameItem.GetComponent<GameItem>();

        return gameItem;
    }

    public IMagic CreateMagic(int nMasterDataID)
    {
        MasterData.Magic masterData = null;
        MasterDataManager.Instance.GetData<MasterData.Magic>(nMasterDataID, ref masterData);

        if (masterData.m_strClassName == "Meteor")
        {
            return new Magic.Meteor();
        }
        else if (masterData.m_strClassName == "WaterJail")
        {
            return new Magic.WaterJail();
        }
        else if (masterData.m_strClassName == "Shield")
        {
            return new Magic.Shield();
        }
        else if (masterData.m_strClassName == "Boost")
        {
            return new Magic.Boost();
        }
        else if (masterData.m_strClassName == "Weight")
        {
            return new Magic.Weight();
        }
        else if (masterData.m_strClassName == "Thunder")
        {
            return new Magic.Thunder();
        }
        else if (masterData.m_strClassName == "Super")
        {
            return new Magic.Super();
        }

        return null;
    }

    public IMagicObject CreateMagicObject(int nMasterDataID)
    {
        MasterData.MagicObject masterData = null;
        MasterDataManager.Instance.GetData<MasterData.MagicObject>(nMasterDataID, ref masterData);

        if (masterData.m_strClassName == "WeightObject")
        {
            return new MagicObject.WeightObject();
        }
		else if (masterData.m_strClassName == "MeteorObject")
        {
            return new MagicObject.MeteorObject();
        }
		else if (masterData.m_strClassName == "WaterJailObject")
        {
            return new MagicObject.WaterJailObject();
        }

        return null;
    }

    //CharacterAI*  CreateCharacterAI(BaeGameRoom* pGameRoom, int nID, int nMasterDataID);
//    public Projectile     CreateProjectile(BaeGameRoom* pGameRoom, int nProjectorID, int nID, int nMasterDataID);
//    public Item           CreateItem(BaeGameRoom* pGameRoom, long long lTime, int nID, int nMasterDataID);
}